### DISCUSSION TITLE
What should be the correct usage of `clim` with volume rendering?

### DISCUSSION THREAD

**Discussion by FelipeCybis** [pyvista/pyvista#4700] [Q&A]
Hi all,

I started using PyVista recently for volume rendering (thanks for the work, btw). I am new to VTK as well, so I treated the `clim` argument of `add_volume` as one would treat `vmin` and `vmax` arguments in matplotlib, in the sense that the data is somewhat independant of the colorbar.

I've seen some work being done on volume rendering and things similar to what I am asking here in a few issues and PRs (https://github.com/pyvista/pyvista/pull/497, https://github.com/pyvista/pyvista/issues/475, https://github.com/pyvista/pyvista/issues/430#issuecomment-548165868)

Looking at the `add_volume` code, I have some doubts about the actual purpose of `clim`.

To reproduce what  I will show:
```python
import pyvista as pv
from pyvista import examples

# Download a volumetric dataset
vol1 = examples.download_knee_full()
vol2 = examples.download_knee_full()

tmp = vol2["SLCImage"]
vol2.clear_data()
# just to have different names for the scalars
vol2["data"] = tmp
```

## 1. If the scalar in the volume is a `np.uint8` (as a few examples in pyvista)  the `clim` is simply a placeholder for the minimum and maximum value in the scalar. In other words it does not change anything in the color dynamics of the scene: 

```python
cpos = [(-381.74, -46.02, 216.54), (74.8305, 89.2905, 100.0), (0.23, 0.072, 0.97)]

pl = pv.Plotter(shape=(1, 2))

pl.add_volume(vol1, cmap="bone", opacity="linear", clim=(0, 255))
pl.camera_position = cpos

pl.subplot(0, 1)
pl.add_volume(vol2, cmap="bone", opacity="linear", clim=(-40, -23))
pl.camera_position = cpos
pl.show()
```
![screenshot](https://github.com/pyvista/pyvista/assets/41338087/313d70bb-194c-40c6-a779-30a8bb2caf96)

## 2. If the scalar is not a `np.uint8`, then the scalar enters in this `if`:

https://github.com/pyvista/pyvista/blob/d09de76fef6a1164eda2f63dfdfb2fd6a75c308a/pyvista/plotting/plotter.py#L4135-L4148

Where there is a few possibilities. The `clim` values can be **inside** the `[min, max]` values of the scalar, it will clip the value with `clim[0]` and `clim[1]` and map everything to `[0, 255]` (the rendering here is what I would expect from `clim`), but if the `clim` values are **outside** of the scalar value range, we can put it wherever, it will always map to `min` or `max`.

```python
cpos = [(-381.74, -46.02, 216.54), (74.8305, 89.2905, 100.0), (0.23, 0.072, 0.97)]

pl = pv.Plotter(shape=(1, 2))

vol1["SLCImage"] = vol1["SLCImage"].astype(float)
pl.add_volume(vol1, cmap="bone", opacity="linear", clim=(0, 50))
pl.camera_position = cpos

vol2["data"] = vol2["data"].astype(float)
pl.subplot(0, 1)
pl.add_volume(vol2, cmap="bone", opacity="linear", clim=(-400, 50))
pl.camera_position = cpos
pl.show()
```

![screenshot1](https://github.com/pyvista/pyvista/assets/41338087/b7408daf-84bc-4a53-bfc5-f1caf37a4977)

To illustrate that, if `clim[0] < min_` and `clim[1] > max_`, this is will be equivalent to `clim=(min_, max_)`) (here, `vol["SLCImage"].min() = 0` and `vol["SLCImage"].max() = 174`.
```python
cpos = [(-381.74, -46.02, 216.54), (74.8305, 89.2905, 100.0), (0.23, 0.072, 0.97)]

pl = pv.Plotter(shape=(1, 2))

vol1["SLCImage"] = vol1["SLCImage"].astype(float)
pl.add_volume(vol1, cmap="bone", opacity="linear", clim=(vol1["SLCImage"].min(), vol1["SLCImage"].max()))
pl.camera_position = cpos

vol2["data"] = vol2["data"].astype(float)
pl.subplot(0, 1)
pl.add_volume(vol2, cmap="bone", opacity="linear", clim=(-400, 500))
pl.camera_position = cpos
pl.show()
```

![screenshot4](https://github.com/pyvista/pyvista/assets/41338087/f24f012d-1109-4281-b02c-33a56acd1174)

## What I would expect

For point 1. I would expect to see the first volume as is, and the second totally saturated at the upper part of the scalar bar (I changed the colorbar just so it has some color, otherwise it will be all white with `"bone"` colormap)

![screenshot2](https://github.com/pyvista/pyvista/assets/41338087/00dc308e-031a-4642-bbe8-a005f6950f59)

For point 2. first plot I would expect the first volume also as is, and the second, since there is data only between 0 and 50, it will also be quite saturated to the upper part of the colormap, but not as much as above.

![screenshot3](https://github.com/pyvista/pyvista/assets/41338087/0de64974-961b-4745-b528-a3acc71a40fc)

For the second plot:
![screenshot5](https://github.com/pyvista/pyvista/assets/41338087/eb7b2b71-8b2b-40c2-9be8-84fec9ddff61)

For my scenes, I've changed the snippet of code I copied just above from [pyvista/pyvista/plotting/plotter.py](https://github.com/pyvista/pyvista/blob/d09de76fef6a1164eda2f63dfdfb2fd6a75c308a/pyvista/plotting/plotter.py#L4135-L4148) for something like:

```python
# data must be between [0, 255], but not necessarily UINT8
# Preserve backwards compatibility and have same behavior as VTK.
# must copy to avoid modifying inplace and remove any VTK weakref
scalars = np.array(scalars, dtype=float)
clim = np.asarray(clim, dtype=scalars.dtype)
scalars.clip(clim[0], clim[1], out=scalars)
if log_scale:
    out = matplotlib.colors.LogNorm(clim[0], clim[1])(scalars)
    scalars = out.data * 255
else:
    np.true_divide((scalars - clim[0]), (clim[1] - clim[0]) / 255, out=scalars, casting='unsafe')
```
I do not have all possible repercussion in my head of what this could "break", especially because casting to float would be needed for correct mapping, but today we cannot plot 2 volumes in the same scene with the same scalar bar, since they will both map their respective range of values to the same scalar bar. Or animating a scene where the scalar bar is the same during the animation and the dynamic changes across time.

Finally, my question is if the behavior we see today is what it should be or could this be an improvement?

**Comment by MatthewFlamm** [comment]
This was also discussed in #4801.  I also think this is confusing behavior.  Importantly, in this discussion and the linked PR, there are situations using `uint8` where the data is not intended to be scaled to [0, 255].  If there needs to be a quick method to get this behavior, Id rather have a parameter that switches it on rather than depending on the type.

But this is likely used by others now, and would be a hard breaking change to find.  We'd have to carefully think about how to deprecate it, if it is changed

**Reply by MatthewFlamm** [reply]
Also, your investigation is quite thorough!  Thanks for putting effort into investigating this.

**Comment by FelipeCybis** [comment]
I understand your concern. I think these discutions reveal that people are using `clim` for two different features.

The first one is `clim` being used as a lookup table to some other range of values, i.e. there is a RGB or `int8` or `[0, 255]` stack of images or data and there is a mapping from this `int8` values to their actual values in some physical unit. This works as intended (modulo if the `max(data)` is not 255). This is not my use case and I don't know if this is common in the 3D rendering community.

The second one is `clim` being used as it is normally used within the 2D plotting libraries in Python. That is, a color scale that will color the data point using a lookup table that links a value to a color.

Note that if you have the first use case, you can map your data beforehand to the physical measure and use `clim` as the second use case with no difference. In PyVista this would be just passing another scalar to the volume.

What I do today is getting the volume with `volume = plotter.add_volume`, and then `volume.mapper.dataset["Data"] = _correct_pyvista_mapping(grid, vmin, vmax)` which is mapping my data to `[0, 255]` using the specified `vmin = clim[0]` and `vmax=clim[1]` and not the `min(data)` and `max(data)`.

Maybe, as you said, having something to switch between behaviors is a good idea. Or explicitly having `vmin, vmax` as `add_volume` arguments and if they are passed, use them for the `[0, 255]` mapping?

**Comment by sampotter** [comment]
Has there been any progress with this? Right now it is very difficult to use `add_volume` to make an animation with a fixed scalar bar.

**Comment by banesullivan** [comment]
FYI, attempting to solve this in #7358
